### PR TITLE
Helper function to retrieve date-sorted filenames for IDF's or Parameter files

### LINKED ISSUE #24797: Helper function to retrieve date-sorted filenames for IDF's or Parameter files
The date-resolving logic within `ExperimentInfo::getInstrumentFilename(instrumentName, date)` can be extracted to a method `ExperimentInfo::resolveDatedFilename(prefix, date)` that will search for file name matching `prefix.*\\.(xml|nxs|hdf5)` within the instrument directories. This method can be used to match IDF and Parameter files:

```
ExperimentInfo::getInstrumentFilename(instrumentName, date){
    return resolveDatedFilename(instrumentName+'_Definition', date)
}

ExperimentInfo::getParameterFilename(instrumentName, date){
    return resolveDatedFilename(instrumentName+'_Parameters', date)
}
```
Possible applications of this in algorithms `LoadInstrument` and `LoadParameterFile`

### PR DESCRIPTION
**Pull request by jmborr**
**Description of work.**
- [x] Helper function `getResourceFilenames` returns a list of IDF's or Parameter files sorted by older `valid-from` dates
- [x] Unit test for `getResourceFilenames`
- [x] Refactor `getInstrumentFilename` to use `getResourceFilenames`
- [x] Python export of `getResourceFilenames`

**To test:**
Download and unzip [instrument.zip](https://github.com/mantidproject/mantid/files/2895477/instrument.zip)  under `/tmp`, then run the following script:
```
instrument_directories = ['/tmp/instrument']
run_number__to__filename_tail = {'BSS_55555': 'Parameters_2011_2018.xml',
                                 'BSS_90177': 'Parameters.xml'}
for run_number, tail in run_number__to__filename_tail.items():
    w = Load(Filename=run, BankName='bank1')
    start_date = str(w.getRun().startTime()).strip()
    parm_files = ExperimentInfo.getResourceFilenames('BASIS_.*Parameters',
                                                     ['xml'],
                                                     instrument_directories,
                                                     start_date)
    for file_path in parm_files:
        assert tail in file_path
```

Fixes #24797

This does not require release notes because of backend changes not exposed to the user

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/IndividualTicketTesting/)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

**Comment by jmborr** [comment]
@AndreiSavici build servers report "squeaky clean"

**Review by AndreiSavici — DISMISSED** [review]

File: Framework/API/src/ExperimentInfo.cpp [hunk 1]
```diff
@@ -1025,25 +1021,77 @@ ExperimentInfo::getInstrumentFilename(const std::string &instrumentName,
...
         }
       }
     }
   }
-  g_log.debug() << "Instrument file selected is " << mostRecentInstFile << '\n';
-  return mostRecentInstFile;
+
+  // Retrieve the file names only
+  std::vector<std::string> pathNames;
+  if (matchingFiles.size() > 0) {
```
Comment (soininen) [inline Framework/API/src/ExperimentInfo.cpp]: The emptyness of a container should be checked using `empty()`. While at it, I propose small modifications to the surrounding code as well:
```c++
  if (!matchingFiles.empty()) {
    pathNames.reserve(matchingFiles.size());
    for (auto elem : matchingFiles)
      pathNames.emplace_back(std::move(elem.second));
  } else {
    pathNames.emplace_back(std::move(mostRecentFile));
  }
```

File: Framework/API/src/ExperimentInfo.cpp [hunk 2]
```diff
@@ -1025,25 +1021,77 @@ ExperimentInfo::getInstrumentFilename(const std::string &instrumentName,
...
+
+  // Get the instrument directories for instrument file search
+  const std::vector<std::string> &directoryNames =
+      Kernel::ConfigService::Instance().getInstrumentDirectories();
+
+  // matching files sorted with newest files coming first
+  const std::vector<std::string> matchingFiles = getResourceFilenames(
+      instrument + "_Definition", validFormats, directoryNames, date);
+  std::string instFile;
+  if (matchingFiles.size() > 0) {
```
Comment (soininen) [inline Framework/API/src/ExperimentInfo.cpp]: `if (!matchingFiles.empty())`

File: Framework/API/src/ExperimentInfo.cpp [hunk 3]
```diff
@@ -1025,25 +1021,77 @@ ExperimentInfo::getInstrumentFilename(const std::string &instrumentName,
...
+ *is found
+ */
+std::string
+ExperimentInfo::getInstrumentFilename(const std::string &instrumentName,
+                                      const std::string &date) {
+  const std::vector<std::string> validFormats = {"xml", "nxs", "hdf5"};
+  g_log.debug() << "Looking for instrument file for " << instrumentName
+                << " that is valid on '" << date << "'\n";
+  // Lookup the instrument (long) name
+  std::string instrument(
```
Comment (soininen) [inline Framework/API/src/ExperimentInfo.cpp]: `const`?

File: Framework/PythonInterface/mantid/api/src/Exports/ExperimentInfo.cpp [hunk 4]
```diff
@@ -42,7 +55,22 @@ void export_ExperimentInfo() {
...
+           "Compile a list of files in compliance with name pattern-matching,\n"
+           "file format, and date-stamp constraints\n\n"
+           "Ideally, the valid-from and valid-to of any valid file should\n"
+           "encapsulate the argument date. If this is not possible, then\n"
+           "the file with the most recent valid-from stamp is selected\n\n"
+           "prefix:         the name of a valid file must begin with this "
+           "pattern\n"
+           "fileFormats:    list of valid file extensions\n"
+           "directoryNames: list of directories to be searched\n"
+           "date :          valid-from and valid-to of a valid file "
```
Comment (soininen) [inline Framework/PythonInterface/mantid/api/src/Exports/ExperimentInfo.cpp]: Nice documentation! Would it make sense to mention the format in which the `date` string should be given?

**Review by soininen — CHANGES_REQUESTED** [review]
A few minor comments below.

File: Framework/PythonInterface/mantid/api/src/Exports/ExperimentInfo.cpp (hunk 4 above)
Comment (jmborr) [reply Framework/PythonInterface/mantid/api/src/Exports/ExperimentInfo.cpp]: https://github.com/mantidproject/mantid/pull/24798/commits/4c2177ad91268fbd7a9bf3a152861b65e78fe9db

File: Framework/API/src/ExperimentInfo.cpp (hunk 3 above)
Comment (jmborr) [reply Framework/API/src/ExperimentInfo.cpp]: https://github.com/mantidproject/mantid/pull/24798/commits/ff6ff6f5c01f36a6c6dc68cd71535b48bf4a6539

File: Framework/API/src/ExperimentInfo.cpp (hunk 2 above)
Comment (jmborr) [reply Framework/API/src/ExperimentInfo.cpp]: https://github.com/mantidproject/mantid/pull/24798/commits/ff6ff6f5c01f36a6c6dc68cd71535b48bf4a6539

File: Framework/API/src/ExperimentInfo.cpp (hunk 1 above)
Comment (jmborr) [reply Framework/API/src/ExperimentInfo.cpp]: https://github.com/mantidproject/mantid/pull/24798/commits/ff6ff6f5c01f36a6c6dc68cd71535b48bf4a6539

**Comment by jmborr** [comment]
@soininen thanks for your comments, I just pushed them

**Review by soininen — APPROVED** [review]

**Comment by jmborr** [comment]
@AndreiSavici @soininen 💃 thanks 💃

**Pull request merged**